hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/DataNodeMetricsService.java

```diff
@@ -0,0 +1,284 @@
+
+    this.reconNodeManager = (ReconNodeManager) reconSCM.getScmNodeManager();
+    this.httpsEnabled = HttpConfig.getHttpPolicy(config).isHttpsEnabled();
+    this.minimumApiDelayMs = config.getTimeDuration(
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY,
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.asyncServiceProvider = metricsServiceProviderFactory.getJmxMetricsServiceProvider();
+  }
+
+  /**
+   * Starts an asynchronous metrics collection task if one is not already running.
+   * Returns immediately - use getCollectedMetrics() to check status and results.
+   */
+  public synchronized void startTask() {
+    MetricsState state = currentState.get();
+    // Check if already running
+    if (state.status == MetricCollectionStatus.IN_PROGRESS) {
+      LOG.warn("Metrics collection task is already in progress");
+      return;
+    }
+    // Check rate limit
+    if (state.lastCollectionEndTime > System.currentTimeMillis() - minimumApiDelayMs) {
+      LOG.info("Skipping metrics collection due to rate limit ({}ms)", minimumApiDelayMs);
+      return;
+    }
+    Set<DatanodeDetails> nodes = reconNodeManager.getNodeStats().keySet();
+    if (nodes.isEmpty()) {
+      LOG.warn("No datanodes found to query");
+      currentState.set(new MetricsState(
+          MetricCollectionStatus.SUCCEEDED,
+          new ArrayList<>(),
+          0L, 0, 0,
+          System.currentTimeMillis()
+      ));
+      return;
+    }
+    LOG.info("Starting async metrics collection for {} datanodes", nodes.size());
+    // Update state to IN_PROGRESS
+    currentState.set(new MetricsState(
+        MetricCollectionStatus.IN_PROGRESS,
+        new ArrayList<>(),
+        0L, 0, 0,
+        state.lastCollectionEndTime
+    ));
+    collectMetricsAsync(nodes);
+  }
+
+  /**
+   * Collects metrics from all datanodes asynchronously using CompletableFuture.
+   */
+  private void collectMetricsAsync(Set<DatanodeDetails> nodes) {
+    int totalNodes = nodes.size();
+
+    // Create futures for all nodes
+    List<CompletableFuture<DatanodePendingDeletionMetrics>> futures = nodes.stream()
+        .map(this::collectFromSingleDataNode)
+        .collect(Collectors.toList());
+
+    // Combine all futures
+    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+        .handle((result, throwable) -> {
+          // Collect all completed results (including failures marked as -1)
+          List<DatanodePendingDeletionMetrics> allResults = new ArrayList<>();
+          long totalPendingDeletion = 0L;
+          int failedCount = 0;
+
+          for (CompletableFuture<DatanodePendingDeletionMetrics> future : futures) {
+            try {
+              // getNow with default will return the result if completed, or default if not
+              DatanodePendingDeletionMetrics metrics = future.get(DEFAULT_TIMEOUT, TimeUnit.SECONDS);
```

Comment:
This timeout is of no use and doesn't work as intended. `allOf()` waits for ALL futures to complete before invoking `handle()`. By the time `future.get(60s)` is called, all futures are already complete, so the timeout parameter is ignored. So, If any single future hangs indefinitely, then `allOf() `waits forever. `handle()` callback is never invoked. `future.get(timeout) `line is never reached. State stays as `IN_PROGRESS` permanently. So it is a permanent service degradation, One stuck network connection can permanently disable the entire metrics collection feature.